### PR TITLE
[tool] Tool supports input_settings

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
Using tool function to configure input settings
- is_multi_select
- allow_manual_entry
- enabled_by
- enabled_by_value
- dynamic_list

```python
from promptflow import InputSettings, DynamicList, tool

@def my_list_func(prefix: str = "", size: int = 10, **kwargs):
    pass


dynamic_list_setting = DynamicList(function=my_list_func, input_mapping={"prefix": "input_prefix"})
input_settings = {
    "input_text": InputSettings(
        dynamic_list=dynamic_list_setting,
        allow_manual_entry=True,
        is_multi_select=True
    )
}

@tool(
    name="My Tool with Dynamic List Input",
    description="This is my tool with dynamic list input",
    input_settings=input_settings
)
def my_tool(input_text: list, input_prefix: str) -> str:
    return f"Hello {input_prefix} {','.join(input_text)}"
```
Generated tool meta:
```json
{
    "name": "My Tool with Dynamic List Input",
    "type": "python",
    "inputs": {
        "input_text": {
            "type": [
                "list"
            ],
            "is_multi_select": true,
            "allow_manual_entry": true,
            "dynamic_list": {
                "func_path": "cli_tool_package.tool_with_dynamic_list_input.my_list_func",
                "func_kwargs": [
                    {
                        "name": "prefix",
                        "type": [
                            "string"
                        ],
                        "reference": "${inputs.input_prefix}",
                        "optional": true,
                        "default": ""
                    },
                    {
                        "name": "size",
                        "type": [
                            "int"
                        ],
                        "optional": true,
                        "default": 10
                    }
                ]
            }
        },
        "input_prefix": {
            "type": [
                "string"
            ]
        }
}
```

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/__init__.py [hunk 1]
```diff
@@ -25,5 +25,7 @@
     "load_run",
     "log_metric",
     "ToolProvider",
+    "InputSettings",
```
Comment (wangchao1230) [inline src/promptflow/promptflow/__init__.py:28]: I feel we should not put it in root namesapce. 

promptflow.entities.InputSettings?

Comment (wangchao1230) [inline src/promptflow/promptflow/__init__.py:28]: InputSettings -> InputSetting

File: src/promptflow/promptflow/_core/tool.py [hunk 2]
```diff
@@ -127,3 +140,56 @@ def get_required_initialize_inputs(cls):
                 if k != "self" and v.default is inspect.Parameter.empty
             }
         return cls._required_initialize_inputs
+
+
+@dataclass
+class DynamicList:
+    function: InitVar[Union[str, Callable]]
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_core/tool.py:147]: add docstrings for this class

File: src/promptflow/promptflow/_core/tool.py [hunk 3]
```diff
@@ -127,3 +140,56 @@ def get_required_initialize_inputs(cls):
...
+                if name in self._input_mapping:
+                    input_info["reference"] = f"${{inputs.{self._input_mapping[name]}}}"
+                input_info["optional"] = value.default is not inspect.Parameter.empty
+                if input_info["optional"]:
+                    input_info["default"] = value.default
+                self.func_kwargs.append(input_info)
+
+
+@dataclass
+class InputSettings:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_core/tool.py:190]: docstring for this class

File: src/promptflow/promptflow/__init__.py (hunk 1 above)
Comment (wangchao1230) [reply src/promptflow/promptflow/__init__.py:28]: maybe put it in promptflow.contracts.tool

File: src/promptflow/promptflow/_utils/tool_utils.py [hunk 4]
```diff
@@ -112,6 +112,8 @@ def function_to_interface(f: Callable, initialize_inputs=None, gen_custom_type_c
     all_inputs = {}
     input_defs = {}
     connection_types = []
+    # Initialize the counter for prompt template
+    prompt_template_count = 0
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_utils/tool_utils.py:116]: why do we need this counter?

File: src/promptflow/promptflow/_core/tool.py [hunk 5]
```diff
@@ -55,11 +59,15 @@ def tool(
     :type description: str
     :param type: The tool type.
     :type type: str
+    :param type: Input setting.
+    :type type: promptflow.InputSettings
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_core/tool.py:63]: dict of InputSetting?

File: src/promptflow/promptflow/__init__.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow/promptflow/__init__.py]: move it to promptflow.entities

Comment (lalala123123) [reply src/promptflow/promptflow/__init__.py]: change to InputSetting

File: src/promptflow/promptflow/_core/tool.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool.py:148]: Add docstring to these classes

File: src/promptflow/promptflow/_core/tool.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool.py]: Add docstring to these classes

File: src/promptflow/promptflow/_core/tool.py (hunk 5 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool.py]: fix it

File: src/promptflow/promptflow/_utils/tool_utils.py (hunk 4 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_utils/tool_utils.py]: remove useless code

**Review by wangchao1230 — DISMISSED** [review]

File: src/promptflow/tests/sdk_cli_test/e2etests/test_tool.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        assert tool_meta == expect_tool_meta
+
+        tool_path = TOOL_ROOT / "tool_with_enabled_by_value.py"
+        tool_meta = self.get_tool_meta(tool_path)
+        expect_tool_meta = {
+            'test_tool.tool_with_enabled_by_value.my_tool': {
+                'name': 'My Tool with Enabled By Value',
+                'type': 'python',
+                'inputs': {
+                    'user_type': {'type': ['string']},
```
Comment (dorisjoy) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_tool.py:186]: The `user_type` doesn't have enum value here ? it should include enum: - student; - teacher

File: src/promptflow/tests/test_configs/tools/tool_with_enabled_by_value.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    STUDENT = "student"
+    TEACHER = "teacher"
+
+
+@tool(
+    name="My Tool with Enabled By Value",
+    description="This is my tool with enabled by value",
+    input_settings={"teacher_id": InputSetting(enabled_by="user_type", enabled_by_value=[UserType.TEACHER])}
+)
+def my_tool(user_type: Enum, student_id: str = "", teacher_id: str = "") -> str:
```
Comment (dorisjoy) [inline src/promptflow/tests/test_configs/tools/tool_with_enabled_by_value.py:17]: Can we change the `student_id` also enabled_by user_type, enabled_by_value is UserType.Student ?

Comment (lalala123123) [reply src/promptflow/tests/test_configs/tools/tool_with_enabled_by_value.py]: Update the test case

File: src/promptflow/tests/sdk_cli_test/e2etests/test_tool.py (hunk 6 above)
Comment (lalala123123) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_tool.py]: Fix it, update test case

**Review by dorisjoy — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**